packages/node-cli/src/commands/actionresult/ActionResult.js

```diff
@@ -0,0 +1,95 @@
+/*
+** Copyright (c) 2019 Oracle and/or its affiliates.  All rights reserved.
+** Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl.
+*/
+'use strict';
+const assert = require('assert');
+
+const ERROR = "ERROR";
+const SUCCESS = "SUCCESS";
+
+class ActionResult {
+
+	constructor(build) {
+		this.validateBuild(build);
+
+		this._status = build.status;
+		this._data = build.data;
+		this._resultMessage = build.resultMessage;
+		this._errorMessages = build.errorMessages;
+	}
+
+	validateBuild(build) {
+		assert(build);
+		assert(build.status, "status is required when creating an ActionResult object.");
+		if (build.status === SUCCESS) {
+			assert(build.data, "data is required when ActionResult is a success.");
+		}
+		if (build.status === ERROR) {
+			assert(build.errorMessages, "errorMessages is required when ActionResult is an error.");
+		}
+	}
+
+	get status() {
+		return this._status;
+	}
+
+	get errorMessages() {
+		return this._errorMessages;
+	}
+
+	get data() {
+		return this._data;
+	}
+
+	get resultMessage() {
+		return this._resultMessage;
+	}
+
+	static get Builder() {
```

Comment:
we could update it to:

```
class ActionResult {
  constructor(build) { ... }
  ...
  static get Builder() {
    return new ActionResultBuilder();
  }
}

class ActionResultBuilder {
  constructor() { }
  withSuccess() {
    this.status = SUCCESS;
    return this;
  }
  ...
  build() { ...}
}
module.exports.ActionResult = ActionResult;
module.exports.ActionResultBuilder = ActionResultBuilder;
```

So that, in the subActionResult that extend ActionResult we can extend as well the ActionResultBuilder from here, like:
```
class CreateProjectActionResult extends ActionResult {
  constructor(build) {
    super(build)
    ...
  }
  ...
  static get Builder() {
    return new CreateProjectActionResultBuilder();
  }
}

class CreateProjectActionResultBuilder extends ActionResultBuilder {
  withProjectType(projectType) {
    this.projectType = projectType;
    return this;
  }
  build() {...}
  ...
}
```

And you don't need to duplicate all the time the methods .withSuccess, withResultMessage, withData, withError, etc

Reply:
I have another similar approach to embed the Builder inside the ActionResult.js
```
class  ActionResult {
	...
	static BuilderClass = class Builder {

		constructor() { }
 
		withSuccess() {
			this.status = SUCCESS;
			return  this;
		}
		...
	}

	static  get  Builder() {
		return  new  this.BuilderClass();
	}
}
```

And then from 
```
class DeployActionResult extends ActionResult {
	...
	static  BuilderClass = class  DeployBuilder  extends  ActionResult.BuilderClass {
		isServerValidation(isServerValidation) {
			this.isServerValidation = isServerValidation;
			return  this;
		}
		appliedContentProtection(appliedContentProtection) {
			this.appliedContentProtection = appliedContentProtection;
			return  this;
		}
	build() {
		return  new  DeployActionResult({
			status:  this.status,
			...(this.data && { data:  this.data }),
			...(this.resultMessage && { resultMessage:  this.resultMessage }),
			...(this.errorMessages && { errorMessages:  this.errorMessages }),
			...(this.isServerValidation && { isServerValidation:  this.isServerValidation }),
			...(this.appliedContentProtection && { appliedContentProtection: this.appliedContentProtection })
		});
	}
}
```

Reply:
I did first approach because I didn't see the second. So, what we do at the end?

Reply:
First option is good!